samcli/lib/telemetry/metric.py

```diff
@@ -29,16 +30,120 @@
+    (by calling flush(), see samcli/cli/main.py)
+
+    Examples
+    --------
+    tracking metric with metric-specific attributes
+    >>> track_metric(MetricName.commandRun, {"foo": "bar"})
+    >>> track_metric(MetricName.commandRun, metric_specific_attrs={"foo": "bar"})
+
+    overwriting exisitng metric-specific attributes
+    (only applicable for metric that can be reported only once, use with caution)
+    >>> track_metric(MetricName.commandRun, {"foo": "bar"}, overwrite_existing_metric_specific_attrs=True)
+
+    tracking metric with top-level attributes
+    >>> track_metric(MetricName.commandRun, top_level_attrs={"foo": "bar"})
+
+    overwriting exisitng top-level attributes
+    (only applicable for metric that can be reported only once, use with caution)
+    >>> track_metric(MetricName.commandRun, top_level_attrs={"foo": "bar"}, overwrite_existing_top_level_attrs=True)
+
+    Parameters
+    ----------
+    metric_name: MetricName
+        an instance of MetricName, e.g. MetricName.commandRun
+    metric_specific_attrs: dict
+        metric specific attributes to track. This is the normally the place where you want to track
+        certain attributes for specific metrics.  e.g. tracking usage of 'use-container' for
+        sam build command.
+    overwrite_existing_metric_specific_attrs: bool
+        If certain attribute is already tracked in metric specific attributes, set this flag to True
+        to overwrite
+    top_level_attrs: dict
+        top level attrs to track. These are the attributes that apply to all metrics.
+    overwrite_existing_top_level_attrs: bool
+        If certain attribute is already tracked in top level attributes, set this flag to True to overwrite
+        Note: normally you don't want to overwrite top level attrs
+    --
+    """
+    metric_sink = _get_global_metric_sink()
+    if metric_name not in metric_sink or not metric_sink[metric_name] or not metric_name.can_report_once_only:
+        metric = Metric(metric_name)
+        metric_sink[metric_name].append(metric)
+    else:
+        metric = metric_sink[metric_name][0]
+
+    metric_specific_attrs = metric_specific_attrs or {}
+    for key, val in metric_specific_attrs.items():
+        if key not in metric["metricSpecificAttributes"] or overwrite_existing_metric_specific_attrs is True:
+            metric["metricSpecificAttributes"][key] = val
+
+    top_level_attrs = top_level_attrs or {}
+    for key, val in top_level_attrs.items():
+        if key not in metric or overwrite_existing_top_level_attrs is True:
+            metric[key] = val
 
 
 def send_installed_metric():
     LOG.debug("Sending Installed Metric")
 
     telemetry = Telemetry()
-    metric = Metric("installed")
-    metric.add_data("osPlatform", platform.system())
-    metric.add_data("telemetryEnabled", bool(GlobalConfig().telemetry_enabled))
+    # NOTE: not using track_metric here because we want to emit the metric right away
```

Comment:
we should note that the concept of a session does not apply here. Good call out.